### PR TITLE
Fixed #34464 -- Fixed queryset aggregation over group by reference.

### PR DESCRIPTION
**Pull request by charettes**
Regression in 59bea9efd2768102fc9d3aedda469502c218e9b7.

Refs ticket-28477.

Thanks Ian Cubitt for the report.

### CODE REVIEW

File: tests/aggregation/tests.py [hunk 1]
```diff
@@ -2175,3 +2176,9 @@ def test_referenced_aggregate_annotation_kept(self):
         sql = ctx.captured_queries[0]["sql"].lower()
         self.assertEqual(sql.count("select"), 2, "Subquery wrapping required")
         self.assertEqual(sql.count("authors_count"), 2)
+
+    def test_referenced_group_by_annotation_kept(self):
+        queryset = Book.objects.values(pages_mod=Mod("pages", 10)).annotate(
+            mod_count=Count("*")
+        )
+        self.assertEqual(queryset.count(), 1)
```
Comment (charettes) [inline tests/aggregation/tests.py:2184]: Query went from

```sql
SELECT COUNT(*)
FROM
  (SELECT "aggregation_book"."id" AS "col1"
   FROM "aggregation_book"
   GROUP BY MOD("aggregation_book"."pages", 10),
            1) subquery;
```

to

```sql
SELECT COUNT(*)
FROM
  (SELECT MOD("aggregation_book"."pages", 10) AS "pages_mod"
   FROM "aggregation_book"
   GROUP BY 1) subquery;
```

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -453,6 +453,9 @@ def get_aggregation(self, using, aggregate_exprs):
                     # filtering against window functions is involved as it
                     # requires complex realising.
                     annotation_mask = set()
+                    if isinstance(self.group_by, tuple):
```
Comment (charettes) [inline django/db/models/sql/query.py:456]: By running the aggregation related tests (`aggregation aggregation_regress queries expressions`) I've confirmed that the only other case where this branch is reached with an empty `annotation_mask` was from `aggregation_regress.tests.AggregationTests.test_more_more_more`

https://github.com/django/django/blob/4bf4222010fd8e413963c6c873e4088614332ef9/tests/aggregation_regress/tests.py#L1071-L1075

Thing is this test group by a column so it ends up in the select clause even if no annotations a selected.

All that to say that we didn't have a single test that covered the case where we group solely by an annotation (not a column) and performed aggregation over it.

**Review by shangxiao — APPROVED** [review]
Nice 👍

I've confirmed the patch works with the reported example + another contrived example I created (watched it go from fail to pass with this branch)

**Comment by felixxm** [comment]
@charettes Thanks :+1:

**Comment by felixxm** [comment]
Merged in 9daf8b4109c3e133eb57349bb44d73cc60c5773c.

**Comment by dtorres-sf** [comment]
@felixxm will this fix be making it in 4.2.X, this regression has caused us issues when recently upgrading to 4.2 (from 3.2) I see it was not in the latest bug fix release 4.2.2, any chance it makes it in the next 4.2 release?

**Comment by charettes** [comment]
@dtorres-sf the patch should be part of Django 4.2.1 which [was released on May 3rd](https://docs.djangoproject.com/en/4.2/releases/4.2.1/#bugfixes).

**Comment by dtorres-sf** [comment]
Thanks, diving deeper it turns out this is not the fix we needed - I posted a comment to the Trac ticket describing instead of here.

**Pull request closed without merging**